rfcs/0010-additional-target-hooks.md

```diff
@@ -0,0 +1,155 @@
+
+TVM_REGISTER_TARGET_KIND("ethos-n", kDLCPU)
+    .set_attr<String>("relay_to_runtime", "relay.ext.ethos-n")
+    .set_attr<String>("constant_updater", "relay.ext.ethos-n.constant_updater");
+```
+
+## Relay -> TIR
+With this change, this path splits, depending on whether you wanted to generate a full `Module` or introduce some specific TIR nodes into the code generation flow; the addition of the `relay_to_tir` hook allows you to write trivial external TIR generators such as calling out to a third party library:
+```python
+@tvm.register_func("target.woofles.lowering")
+def tir_generator(ir_module, relay_func):
+    """A simple TIR generator for testing"""
+    ib = tvm.tir.ir_builder.create()
+    A = tvm.tir.decl_buffer(shape=(8,8,), dtype=relay_func.params[0].checked_type.dtype)
+    B = tvm.tir.decl_buffer(shape=(8,8,), dtype=relay_func.params[0].checked_type.dtype)
+    C = tvm.tir.decl_buffer(shape=(8,8,), dtype=relay_func.params[0].checked_type.dtype)
+    ib.emit(
+        tvm.tir.call_extern('int32', 'woofles', A.data, B.data, 8, 8, C.data)
+    )
+
+    prim_func = tvm.tir.PrimFunc([A, B, C], ib.get())
+    new_module = tvm.lower(prim_func, name=relay_func.attrs["global_symbol"])
+
+    return new_module, GlobalVar(relay_func.attrs["global_symbol"])
+```
+This is then registered on a target:
+```
+TVM_REGISTER_TARGET_KIND("woofles", kDLCPU)
+    .set_attr<String>("relay_to_tir", "target.woofles.lowering");
+```
+The signature for this hook is as follows:
+```
+relay_to_tir(const IRModule& ir_module, const relay::Function& function) -> (IRModule, GlobalVar)
+```
+Which takes a read only `IRModule` and relevant `Function` and returns a new `IRModule` which represents the transformed function, alongside a `GlobalVar` which indicates the top-level operator function within that new `IRModule`.
+
+## TIR -> Runtime
+Extending from the above, a second hook is introduced to do further transformations from TIR -> Runtime named `tir_to_runtime`, this bypasses the default `target.build.X` and instead uses the registered `tir_to_runtime` build:
+```
+runtime::Module BuildWooflesHost(IRModule mod, Target target) {
+// ... Custom Code generation here
+}
+
+TVM_REGISTER_GLOBAL("target.build.woofles").set_body_typed(BuildWooflesHost);
+TVM_REGISTER_TARGET_KIND("woofles", kDLCPU)
+    .set_attr<String>("tir_to_runtime", "target.build.woofles");
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This functionality is an extension of the existing use of `attr::kCompiler` to provide a hint that we can use to lookup attached target attribute, the compile engine and code generation flows can choose to store TIR and/or generate runtime modules based on the registered hooks.
+
+## Relay to TIR Hook
+[relay-to-tir-hook]: #relay-to-tir-hook
+
+This can be added to the TE Compiler by cross referencing the existing `attr::kCompiler` with the `TargetKind` registry:
```

Comment:
i guess one dumb question here @jroesch @mbs-octoml: if we are making a generic interface to lowering Relay to TIR, why shouldn't AutoTVM schedule-based lowering also fit this interface? then all the test infra would be available to external codegen folks, which means we would have a much better chance writing a set of asserts to validate lowered TIR

Reply:
That sounds sensible. I need to pin Junru down for an hour or so to get a tour of AutoTVM to understand how subtle the extension points would need to be.